### DISCUSSION TITLE
Allow multiple env_prefix in for BaseSettings

### DISCUSSION THREAD

**Discussion by timbmg** [pydantic/pydantic#4319] [Feature Request]
When using BaseSettings, automatically loading variables from the env is very convenient. I use it, particularly in combination with docker containers. However, the env variables that these services have built-in might have multiple different prefixes. In the example below, RabbitMQ uses the prefix `RABBITMQ_DEFAULT_` (for username and password) and `RABBITMQ_` (for ssl_verify). 

```python
from pydantic import BaseSettings, Field, SecretStr

class RabbitMQSettings(BaseSettings):
    class Config:
        env_prefix = "RABBITMQ_"

    user: str = Field(..., env="RABBITMQ_DEFAULT_USER")
    password: SecretStr = Field(..., env="RABBITMQ_DEFAULT_PASS")
    ssl_verify: str = Field(..., env="RABBITMQ_SSL_VERIFY")
```

It would be great if multiple `env_prefix` could be supported. Then the loading could be completely automatic.

It could like like this:
```python
from pydantic import BaseSettings, Field, SecretStr

class RabbitMQSettings(BaseSettings):
    class Config:
        env_prefix = ["RABBITMQ_", "RABBITMQ_DEFAULT_"]

    user: str 
    password: SecretStr 
    ssl_verify: str
```

**Comment by heymoonman** [comment]
A use case I was looking at that this would help with is having a few general env variables that set username and password for multiple components/containers, but still allowing them to be overridden by a specific component by using that component's specific env_prefix. +1 to this feature.

**Comment by matutter** [comment]
I came accross this today and found a little workaround using the new Pydantic v2 `EnvSettingsSource`. 

```python
from typing import Any, List, Tuple, Type
from pydantic.fields import FieldInfo
from pydantic_settings import (
    BaseSettings,
    EnvSettingsSource,
    PydanticBaseSettingsSource,
)

class ExtendedEnvSettingsSource(EnvSettingsSource):
    def get_field_value(self, field: FieldInfo, field_name: str) -> tuple[Any, str, bool]:
        if prefixes := self.config.get("env_prefixes"):
            for prefix in prefixes:
                self.env_prefix = prefix
                env_val, field_key, value_is_complex = super().get_field_value(field, field_name)
                if env_val is not None:
                    return env_val, field_key, value_is_complex

        return super().get_field_value(field, field_name)


class ExtendedSettingsConfigDict(SettingsConfigDict, total=False):
    env_prefixes: List[str] | None


class ExtendedBaseSettings(BaseSettings):

    @classmethod
    def settings_customise_sources(
        cls,
        settings_cls: Type[BaseSettings],
        init_settings: PydanticBaseSettingsSource,
        env_settings: PydanticBaseSettingsSource,
        dotenv_settings: PydanticBaseSettingsSource,
        file_secret_settings: PydanticBaseSettingsSource,
    ) -> Tuple[PydanticBaseSettingsSource, ...]:
        return (ExtendedEnvSettingsSource(settings_cls),)

class DebugSettings(ExtendedBaseSettings):
    model_config = ExtendedSettingsConfigDict(
        env_prefixes=("debug_", "nomad_meta_debug_"),
    )

    enabled: bool = False
    wait_for_client: bool = False
    port: int = 5678

```

Sometimes my environment is `DEBUG_ENABLED`, usually locally, and then `NOMAD_META_DEBUG_ENABLED` if I am attaching my debugger to a service inside a nomad job.

Reference: https://docs.pydantic.dev/latest/usage/pydantic_settings/#parsing-environment-variable-values

**Reply by Jakub-Ner** [reply]
I found that this part doesn't work for nested envs like `DEBUG_MODEL__NAME`:
> class ExtendedEnvSettingsSource(EnvSettingsSource):
>     def get_field_value(self, field: FieldInfo, field_name: str) -> tuple[Any, str, bool]:
>         if prefixes := self.config.get("env_prefixes"):
>             for prefix in prefixes:
>                 self.env_prefix = prefix
>                 env_val, field_key, value_is_complex = super().get_field_value(field, field_name)
>                 if env_val is not None:
>                     return env_val, field_key, value_is_complex
>         return super().get_field_value(field, field_name)

I recommend using this:
```python
class ExtendedEnvSettingsSource(EnvSettingsSource):
    def __call__(self) -> dict[str, Any]:
        if prefixes := self.config.get("env_prefixes"):
            d = {}
            for prefix in prefixes:
                self.env_prefix = prefix
                d.update(super().__call__())
            return d
        return super().__call__()

````

**Comment by namoshizun** [comment]
+1 This would be a really handy feature for dealing with environment name changes.  Sometimes we would like to rename an environment variable required by a package, but still want to able to read from the old names if they are available, just for backwards compatibility.
